박스오피스 [STEP 2] Volga

volgaHwang opened the pull request:
@zdodev 
안녕하세요 소대!

박스오피스 STEP2 PR을 요청드립니다~

## 고민한 점
이번 STEP에서는 코드 작성 할 때, 많은 고민을 했습니다..!
- `URLSessionNetworkService`라는 Class와 `URLSessionNetworkServiceProtocol`을 만들었습니다.
    - 싱글턴을 만들어서 사용하는 것은 쉬운 방법이지만, 테스팅에 있어서 초기화나 세팅이 어렵다고 판단해서 인스턴스를 주입하는 것을 고려했습니다.
    - 이 문제에서는, 최대한 ViewController에 `URLSessionNetworkService` 클래스에 대한 의존성을 줄여야겠다고 판단을 했고, Protocol을 만들어 놓고 그에 대해 사용하고자 해당 라인을 작성했습니다.
- URLSessionNetworkService 클래스 인스턴스 메소드 fetch를 범용성을 고려하여 Generic을 사용해서 코드를 구현했습니다.
    - completion escaping closure를 통해 데이터 통신 결과물인 Result<T, V> 형태로 뱉게되는데, 이 경우 V에 해당하는 Error는 네트워크 통신에 대한 결과로서 일반적인 추상화가 되지만, T에 해당하는 부분은 그렇지가 않게 됩니다.
    - `Generic`을 사용 할 경우, 받아올 데이터 타입은 사용 할 때 지정해주면 따로 문제가 없다고 판단을 했고 코드 추상화에 도움이 된다고 생각해서 반영하게 됐습니다.
- `URLSessionNetworkService`에서 사용하는 `enumerate`들에 대해 Constant 디렉터리에 따로 추가해서 반영했습니다 (`QueryKeys`, `URLInfo`)
- `Network` 통신에 대한 동작 확인을 위해 Unit Test를 진행했습니다. Completion 이후에 발생하는 비동기적인 일에 대해 테스팅을 진행해야 했는데, 일반적인 블로킹 코드로 작성하게 될 경우에는 URLSession이 네트워크 통신에 대한 모든 동작을 하기 전에 함수를 끝내버리기 때문에 확인이 불가능했습니다. 이 부분에 대해 문제를 해결하고자 `XCTestExpectation`을 사용해서 구현했습니다.
- APP Key를 가지고있어야했는데, 이 부분에 대해서는 info.plist에 담아서 쓰는 것이 좋겠다고 생각했습니다.

## 궁금한 점
- `URLSessionNetworkService`라는 네이밍을 사용하긴 했습니다만, 더 좋은 아이디어가 떠오르지 않네요.. 좋은 아이디어가 있을까요? ㅠㅠ..
- 그리고, Generic을 사용해서 `fetch`메소드에 대한 틀을 구현하긴 했습니다만, 여전히 Decoding을 Completion단계에서 해주는 것이 옳은건지? 에 대한 궁금한 점이 있습니다.
- 그리고 Network 통신에 대한 테스트를 진행하면서 XCTestExpectation을 사용해서 내용물도 프린트 해보고 테스팅을 진행하긴 했지만, 이게 무조건 맞아! 라는 생각은 들진 않는것 같습니다. 혹시 이 방법 외에 비동기 관련한 테스팅 방법론이 있을까요??

감사합니다!

volgaHwang committed:
feat: MovieInfoDetailResult 정의

- Encoding이 필요하지 않아 전부 Decodable Protocol을 채택하는 구조체를 만들었습니다.
- 고민이었던 부분은 과연 nil로 들어오는 부분이 없을 것 인가? 에 대한 의문점인데, 이 부분은 API 명세서에 따로 제시가 되어있지 않으므로, nil값이 들어오지 않는다고 판단했습니다.
- 구조체 이름과 중복된 이름을 사용하는 프로퍼티를 없애기 위해 노력했습니다.

volgaHwang committed:
chore: API 통신에 필요한 AppKey Info.plist에 추가 반영

volgaHwang committed:
feat: URLSession 통신을 위한 API Protocol & Class 생성

- `URLSessionNetworkServiceProtocol` 을 구현했습니다. Test나 나중에 Network Class에 대해 의존성을 줄이는 방향으로 개발을 진행하고자 Protocol을 채택하게 했습니다.
- `URLSessionNetworkService`, `URLSessionNetworkServiceProtocol` 에 대해 Generic을 사용했고, queryItems를 parameter로 받아 fetch 메소드를 추상화 했습니다.
    - URLInfo Enumerate를 통해 API 통신에 필요한 URL 주소를 String으로 가져오게 구현했습니다.
    - queryItems는 Dictionary 형식으로 받아오지만, URLComponentGenerator를 통해 URLComponents에 담길 URLQueryItems로 변환하게 됩니다.
- API 에러처리를 위해 `URLSessionNetworkServiceError` 구현했습니다.
- `Date`에서 오늘 날짜를 가져오기 위해 Extension을 사용하여 오늘 날자를 static computed Property로 가져 올 수 있게 구현했습니다.
- `NetworkTests` 테스트 파일을 생성하여 네트워크 통신 동작에 대한 테스트를 진행했습니다.

volgaHwang committed:
feat: URLSessionNetworkService 프로퍼티 추가 반영

URLSession -> session property 추가 반영

zdodev commented on BoxOffice/Constant/URLInfo.swift:
```diff
@@ -0,0 +1,21 @@
+enum URLInfo {
+    case searchDailyBoxOffice
+    case searchDetailMovieInfo
+    var urlString: String {
+        switch self {
+        case .searchDailyBoxOffice:
+            return "http://kobis.or.kr/kobisopenapi/webservice/rest/boxoffice/searchDailyBoxOfficeList.json?"
+        case .searchDetailMovieInfo:
+            return "http://www.kobis.or.kr/kobisopenapi/webservice/rest/movie/searchMovieInfo.json?"
+        }
+    }
```
영화사 API를 보면 두 URL 동일하게 www.kobis.or.kr~.. 로 시작하고 있습니다!
일별 박스오피스 URL과 영화 상세정보 URL 의 공통의 host(baseURL)를 두고 다른 path 만 별도로 선언하는 것은 어떨까요?

volgaHwang replied:
아주 좋은 생각인것 같습니다. 변경해서 사용해볼게요!

zdodev commented on BoxOffice/Models/MovieListResult.swift:
```diff
@@ -0,0 +1,152 @@
+    var movieInfoResult: MovieInfoResult
+    
+    enum CodingKeys: String, CodingKey {
+        case movieInfoResult = "movieInfoResult"
```
프로퍼티명과 파라미터명이 동일한 부분에도 `CodingKey` 를 적용하신 이유가 궁금합니다!

volgaHwang replied:
특별한 이유는 없었고 관용적으로 넣어야 되는거 아닌가? 하는 생각에 넣은것 같습니다.
빼볼게요!

zdodev commented on BoxOffice/Network/Protocol/URLSessionNetworkServiceProtocol.swift:
```diff
@@ -0,0 +1,16 @@
+
+import Foundation
+
+protocol URLSessionNetworkServiceProtocol {
```
네트워크 `fetch` 에 대한 프로토콜을 구현해보셨네요.👍

zdodev commented on BoxOffice/Network/URLSessionNetworkService.swift:
```diff
@@ -0,0 +1,84 @@
+import Foundation
+
+final class URLSessionNetworkService: URLSessionNetworkServiceProtocol {
+    let session: URLSession
```
`session` 을 `internal` 수준으로 정의하신 이유가 있을까요?

volgaHwang replied:
엇.. private로 표현 되야 하는것인데, 실수로 빼먹은 것 같습니다ㅎㅎ;

zdodev commented on BoxOffice/Network/URLSessionNetworkService.swift:
```diff
@@ -0,0 +1,84 @@
+        self.session = session
+    }
+    
+    private enum AppInfo {
```
`.plist` 파일 읽기 기능을 구현해보셨네요.👍
해당 기능을 `NetworkService` 에서 구현하신 이유가 궁금합니다!

volgaHwang replied:
AppInfo에서 Info.plist에 대한 정보를 가져오는 역할을 URLSessionNertworkService 클래스 내부에서만 필요한 정보고 그렇기에 해당 영역에서만 가져오면 되는 정보라고 생각해서 해당 영역에 넣게 되었습니다.

구현 변경사항 에서는 URLInfo에서 가져올수 있도록 구현했습니다.

zdodev commented on BoxOffice/Network/URLSessionNetworkService.swift:
```diff
@@ -0,0 +1,84 @@
+            }
+            do {
+                let decodedData = try JSONDecoder().decode(T.self, from: data)
+                completion(.success(decodedData))
```
해당 부분은 디코딩하는 부분까지 책임까지 지을 것인가에 대한 문제일 것 같습니다.
디코딩까지 완료해야 성공으로 판단할지 고민해보면 좋을 것 같습니다!

zdodev commented on BoxOffice/Network/URLSessionNetworkService.swift:
```diff
@@ -0,0 +1,84 @@
+        task.resume()
+    }
+    
+    private func URLComponentsGenerator(searchTarget: URLInfo,
```
다음처럼 메서드 이름을 지으신 다른 이유가 있을까요? 아니면 메서드 이름을 API 명명 규칙처럼 변경해보는 것은 어떨까요?

volgaHwang replied:
넵 네이밍 컨벤션을 반영해서 동사를 사용해서 변경해보겠습니다!

zdodev commented on BoxOffice/Network/URLSessionNetworkService.swift:
```diff
@@ -0,0 +1,84 @@
+    
+    private func URLComponentsGenerator(searchTarget: URLInfo,
+                                        queryItems: [String: String]?) -> URLComponents? {
+        var components = URLComponents(string: searchTarget.urlString)
+        components?.queryItems?.append(AppInfo.key)
```
`APPKey` 파라미터 추가와 `queryItems` 파라미터 추가 부분을 분리해보는 것은 어떨까요? 의견이 궁금합니다!

volgaHwang replied:
이 부분에 대해서 고민을 좀 했는데.. 어떠한 식의 분리를 말씀하시는 걸까요? 제가 이해를 잘 못한거같아요 ㅠㅠ

zdodev replied:
전달받은 `queryItems` 파라미터를 추가하는 부분과 `APPKey` 파라미터를 추가하는 부분이 한 메서드 내에서 동작한다고 생각하여 이 두 역할을 각각의 메서드로 분리하면 어떨까 싶었습니다!

volgaHwang replied:
제대로 한게 맞는진 모르겠지만.. 일단 분리는 해봤습니다..!

zdodev commented on BoxOffice/Network/URLSessionNetworkServiceError.swift:
```diff
@@ -0,0 +1,36 @@
+
+import Foundation
+
+enum URLSessionNetworkServiceError: Error {
```
`Error` 프로토콜을 사용하여 다양한 에러 케이스를 정의해보셨네요.👍

zdodev commented on BoxOfficeTests/NetworkTests.swift:
```diff
@@ -0,0 +1,56 @@
+                         queryItems: [QueryKeys.movieCode: "20124079"]) {
+            (networkResult: Result<MovieInfoDetailResult, URLSessionNetworkServiceError>) -> Void in
+            switch networkResult {
+            case .success(let success):
+                expects.fulfill()
```
응답으로 받은 데이터를 검증하는 코드를 추가하는 것은 어떨까요?

zdodev commented on BoxOfficeTests/NetworkTests.swift:
```diff
@@ -0,0 +1,56 @@
+                         queryItems: [QueryKeys.targetDate: Date.today]) {
+            (networkResult: Result<BoxOfficeSearchResult, URLSessionNetworkServiceError>) -> Void in
+            switch networkResult {
+            case .success(let success):
+                print(success)
+                expects.fulfill()
```
`print` 로 출력하기 보다는 응답으로 받은 데이터를 검증하는 코드를 추가하는 것은 어떨까요?

zdodev reviewed (changes requested):
안녕하세요. @volgaHwang !

Step2 잘 구현해주셨습니다!
프로토콜을 사용하여 의존성을 줄일 수 있도록 구현하신 부분도 잘 구현해주셨습니다.👍

질문에 대한 제 의견입니다.
❓URLSessionNetworkService라는 네이밍을 사용하긴 했습니다만, 더 좋은 아이디어가 떠오르지 않네요.. 좋은 아이디어가 있을까요? ㅠㅠ..
- 지금은 네트워크 타입들에 대해 `URLSession` 을 사용하지 않는 정도면 괜찮을 것 같습니다!

❓그리고, Generic을 사용해서 fetch메소드에 대한 틀을 구현하긴 했습니다만, 여전히 Decoding을 Completion단계에서 해주는 것이 옳은건지? 에 대한 궁금한 점이 있습니다.
- Decoding 책임을 어디에 맡기는 것이 좋을지 고민해보면 좋을 것 같습니다! Completion 단계에서 수행하는 것도 하나의 방법이 되지 않을까 싶습니다!

❓그리고 Network 통신에 대한 테스트를 진행하면서 XCTestExpectation을 사용해서 내용물도 프린트 해보고 테스팅을 진행하긴 했지만, 이게 무조건 맞아! 라는 생각은 들진 않는것 같습니다. 혹시 이 방법 외에 비동기 관련한 테스팅 방법론이 있을까요??
- 네트워킹 테스트도 잘 작성해주셨습니다.👍 네트워크에 의존하지 않는 테스트에 대한 방법을 더 알아보시면 좋을 것 같습니다! 동기, 비동기 테스트에 대한 이해도 높아질 것 같습니다!

몇 가지 궁금증을 남겨보았습니다!
확인 부탁드립니다!

volgaHwang committed:
refactor: URLSessionNetworkService session 접근 제한자 수정사항 반영ㅁ

volgaHwang committed:
refactor: MovieInfoDetailResult에서 불필요한 CodingKeys 삭제 반영

volgaHwang committed:
refactor: URLInfo 에 대한 구조체 변경

- URLInfo에서 AppKey를 받아오도록 구현했습니다.
- http://kobis.or.kr에 대한 공통적인 baseurl과, path를 나누었습니다.

volgaHwang committed:
refactor: URLSessionNetworkService 변경사항 반영

- URLComponentsGenerator -> establishURLComponents로 네이밍 변경
- establishURLComponents 내부 메소드 로직 변경사항 반영

volgaHwang committed:
refactor: URLSessionNetworkService: establishURLComponents 쿼리 추가 부분 반영

volgaHwang committed:
test: NetworkTests XCT관련 테스팅 코드 변경사항 반영

- print를 제거해서 테스트를 확인 할 수 있게끔 변경했습니다.

volgaHwang commented:
@zdodev 
안녕하세요 소대~
말씀주신 부분들에 대해 변경사항을 반영했습니다.

질문사항이 있어서 추가적으로 멘션드립니다!
## 질문사항

> `APPKey` 파라미터 추가와 `queryItems` 파라미터 추가 부분을 분리해보는 것은 어떨까요? 의견이 궁금합니다!

- 이 부분에 대해, 어떠한식으로 분리를 하는 것인지에 대해 맥락파악이 잘 안되는 것 같아서 질문을 한번 더 드립니다. 어떠한식의 분리를 말씀하시는걸까요?

volgaHwang committed:
test: Test Double을 위한 Mock 객체 반영

- 호출에 대한 기대를 명세하고, 구현 내용에 대한 테스트를 진행했습니다.
- 추가적으로 기존 테스트를 Network 동작을 가정하고 했었는데, Mock을 이용한 테스트와는 분리를 했습니다.

volgaHwang committed:
refactor: URLSessionNetworkService 리네이밍

- URLSessioNetworkService -> NetworkService
- URLSessioNetworkServiceProtocol -> NetworkServiceProtocol

volgaHwang committed:
refactor: URLSessionNetworkServiceError 리네이밍 반영

volgaHwang committed:
refactor: Query 생성 부분에 대한 함수 분리

- 기존에 AppKey + Query를 한 함수에서 URLQueryItem을 따로 생성해서 합쳤는데, Dict -> URLQueryItem으로 만드는 부분을 함수화 해서 분할했습니다.

volgaHwang committed:
refactor: NetworkService Components 추가 부분 변경사항 반영

volgaHwang committed:
test: NetworkMockTests에서 url 상수 부분으로 변경

volgaHwang commented:
@zdodev 

안녕하세요~ 소대!

연휴 잘보내셨을까요?

오늘의 PR 올립니다.

- 코멘트 주신 사항들에 대해 반영한 패치를 올렸습니다.
- AppKey와 QueryItems를 분할하는 것에 대해서 반영을했습니다.
- 테스트 관련해서 Mock을 만들어서 테스트했습니다.

zdodev commented on BoxOffice/Constant/URLInfo.swift:
```diff
@@ -0,0 +1,36 @@
+import Foundation
+
+enum URLInfo {
+    static let baseURL = "http://www.kobis.or.kr/"
```
좋습니다!👍

zdodev commented on BoxOfficeTests/NetworkTest/MockURLSessionProtocol.swift:
```diff
@@ -0,0 +1,46 @@
+
+typealias RequestURL = String
+
+class MockURLProtocol: URLProtocol {
```
`URLProtocol` 를 사용하여 테스트 코드를 작성해보셨네요.👍

zdodev reviewed (approved):
안녕하세요. @volgaHwang !

코멘트에 대한 의견 잘 보았습니다!

Step 2 구현하시느라 고생하셨습니다!👍
